magefile.go

```diff
@@ -0,0 +1,256 @@
+func Cover() error {
+	mg.Deps(Test)
+	fmt.Println("Running coverage...")
+	return sh.RunV("go", "tool", "cover", "-html=coverage.txt")
+}
+
+// Fmt formats code
+func Fmt() error {
+	fmt.Println("Formatting...")
+	files, err := findFilesRecursive(func(path string, _ os.FileInfo) bool {
+		// ignore go files in /rpc
+		return filepath.Ext(path) == ".go" && !filepath.HasPrefix(path, "rpc/")
+	})
+	if err != nil {
+		return fmt.Errorf("failed to find files: %w", err)
+	}
+
+	args := append([]string{"-w"}, files...)
+	return sh.RunV("goimports", args...)
+}
+
+// Lint runs the linters
+func Lint() error {
+	fmt.Println("Linting...")
+
+	if err := sh.RunV("golangci-lint", "run"); err != nil {
+		return fmt.Errorf("failed to lint: %w", err)
+	}
+
+	return sh.RunV("buf", "lint")
+}
+
+// Prep prepares the project for building
+func Prep() error {
+	fmt.Println("Preparing...")
+	mg.Deps(Clean)
+	mg.Deps(UI.Build)
+	return nil
+}
+
+// Proto generates protobuf files and gRPC stubs
+func Proto() error {
+	fmt.Println("Generating proto files...")
+	return sh.RunV("buf", "generate")
+}
+
+// Test runs the tests
+func Test() error {
+	fmt.Println("Testing...")
+
+	env := map[string]string{
+		"FLIPT_TEST_DATABASE_PROTOCOL": "sqlite3",
+	}
+
+	if os.Getenv("FLIPT_TEST_DATABASE_PROTOCOL") != "" {
+		env["FLIPT_TEST_DATABASE_PROTOCOL"] = os.Getenv("FLIPT_TEST_DATABASE_PROTOCOL")
+	}
+
+	return sh.RunWithV(env, "go", "test", "-v", "-covermode=atomic", "-count=1", "-coverprofile=coverage.txt", "-timeout=60s", "./...")
+}
+
+type UI mg.Namespace
+
+// Build generates UI assets
+func (u UI) Build() error {
+	mg.Deps(u.Deps)
+	fmt.Println("Generating assets...")
+
+	if err := os.Chdir(".build/ui"); err != nil {
+		return fmt.Errorf("failed to change dir: %w", err)
+	}
+
+	defer os.Chdir("../..")
+
+	if err := sh.RunV("npm", "run", "build"); err != nil {
+		return err
+	}
+
+	return sh.RunV("mv", "dist", "../../ui")
+}
+
+// Clone clones the UI repo
+func (u UI) Clone() error {
+	if err := os.MkdirAll(".build", 0755); err != nil {
+		return fmt.Errorf("failed to create dir: %w", err)
+	}
+
+	if _, err := os.Stat(".build/ui/.git"); os.IsNotExist(err) {
+		fmt.Println("Cloning UI repo...")
+		if err := os.Chdir(".build"); err != nil {
+			return fmt.Errorf("failed to change dir: %w", err)
+		}
+
+		defer os.Chdir("..")
+
+		if err := sh.RunV("git", "clone", "https://github.com/flipt-io/flipt-ui.git", "ui", "--depth=1"); err != nil {
+			return fmt.Errorf("failed to clone UI repo: %w", err)
+		}
+	}
+
+	return nil
+}
+
+// Sync syncs the UI repo
+func (u UI) Sync() error {
+	mg.Deps(u.Clone)
+	fmt.Println("Syncing UI repo...")
+
+	if err := os.Chdir(".build/ui"); err != nil {
+		return fmt.Errorf("failed to change dir: %w", err)
+	}
+	defer os.Chdir("../..")
+
+	if err := sh.RunV("git", "fetch", "--all"); err != nil {
+		return fmt.Errorf("failed to fetch UI repo: %w", err)
+	}
+	if err := sh.RunV("git", "reset", "--hard", "origin/main"); err != nil {
```

Comment:
I wonder if it's worth squeezing a checkout to main in here.

Just in case, for some odd reason, something apart from `main` has since been checked out.
Otherwise, that local ref will be updated to point to `main`.